File: pathwaysfinder/api/app/schemas/parts.py
from datetime import datetime
from pydantic import BaseModel


class PartBase(BaseModel):
    name: str
    type: str
    description: str | None = None
    sequence: str
    organism: str | None = None
    source: str | None = None


class PartCreate(PartBase):
    pass


class PartResponse(PartBase):
    id: str
    created_at: datetime
    updated_at: datetime | None = None

    class Config:
        from_attributes = True


class PartListResponse(BaseModel):
    parts: list[PartResponse]
    total: int


class PaperResponse(BaseModel):
    pmid: str | None = None
    title: str | None = None
    authors: list[str] = []
    abstract: str | None = None
    journal: str | None = None
    year: str | None = None
    doi: str | None = None
    doi_url: str | None = None
    url: str | None = None


class PapersListResponse(BaseModel):
    papers: list[PaperResponse]
    query: str

File: pathwaysfinder/frontend/src/types/parts.ts
export interface Part {
  id: string
  name: string
  type: 'promoter' | 'rbs' | 'terminator' | 'gene'
  description: string | null
  sequence: string
  organism: string | null
  source: string | null
  created_at: string
  updated_at: string | null
}

export interface PartListResponse {
  parts: Part[]
  total: number
}

export type PartType = Part['type']

export const PART_TYPES: PartType[] = ['promoter', 'rbs', 'terminator', 'gene']

export const ORGANISMS = ['ecoli', 'yeast'] as const
export type Organism = (typeof ORGANISMS)[number]

export interface Paper {
  pmid: string | null
  title: string | null
  authors: string[]
  abstract: string | null
  journal: string | null
  year: string | null
  doi: string | null
  doi_url: string | null
  url: string | null
}

export interface PapersResponse {
  papers: Paper[]
  query: string
}
